enhancements/machine-api/aws-placement-groups.md

```diff
@@ -0,0 +1,487 @@
+    part of the installation configuration before manifest generation
+- Support for Placement Groups for Control Plane Machines
+  - Users may replace their control plane at a later date using future automation if they wish to have their control
+    plane within a placement group
+- Support for moving Machines between placement groups
+
+## Proposal
+
+We will introduce new fields within the `Placement` section of the AWS Machine Provider Config which will allow users
+to specify which placement group to create instances within.
+
+If the placement group does not exist, it will be created based upon the group type and, if applicable partition count provided.
+
+We will also allow end users to specify which partition their instances should be created within when appropriate.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to use a spread placement group to ensure that my critical
+  workloads are unlikely to suffer from hardware failures simultaneously.
+- As an operator of High-Performance Computing workloads on OpenShift, I would like to use a cluster placement group to
+  ensure that there is
+  minimal latency between instances within my cluster, allowing maximal throughput of my workloads.
+- As an operator of a large scale distributed and replicated data store, I would like to use a partition placement
+  group to ensure that different shards of my data are replicated within different fault domains within the same
+  availability zone.
+- As an OpenShift cluster administrator, I would like unneeded placement groups to be cleaned up so that I am not
+  polluting my environment with resources that are no longer in use.
+
+### API Extensions
+
+The new fields for this enhancement will be limited to the `Placement` struct within the `AWSMachineProviderConfig` in
+the Machine v1beta1 API group.
+
+The changes are outlined below:
+
+```golang
+type Placement struct {
+  // Existing Fields
+  ...
+
+  // GroupName specifies the name of an AWS placement group to create the Machine within.
+  // If the group specified does not exist, it will be created based on the GroupType
+  // paramenter.
+  // When a Machine is deleted, if this leaves an empty Placement Group, the group is also
+  // deleted.
+  // +optional
+  GroupName string `json:"groupName,omitempty"`
+
+  // GroupType specifies the type of AWS placement group to use for this Machine.
+  // This parameter is only used when a Machine is being created and the named
+  // placement group does not exist.
+  // Valid values are "Cluster", "Partition", "Spread", and omitted, which means
+  // no opinion and the platform chooses a good default which may change over time.
+  // The current default value is "Spread".
+  // +kubebuilder:validation:Enum:="Cluster";"Partition";"Spread"
+  // +optional
+  GroupType AWSPlacementGroupType `json:"groupType,omitempty"`
+
+  // PartitionCount specifies the number of partitions for a Partition placement
```

Comment:
This description really makes this sound like it should be part of  discriminated union.

Reply:
So I've added the type as an explicit option so we know what the customer wants us to do, as there (currently) is no configuration apart from for partition types, I didn't think we would need to add another struct level to make it a proper discriminated union, but I guess there's no harm in it

To clarify you're suggestion is:

```
type Placement struct {
  Region
  AvailabilityZone
  Tenancy
  GroupType AWSPlacementGroupType
  GroupName string
  Partition *AWSPartitionPlacement
}

type AWSPartitionPlacement struct {
  Count int32
  Number int32
}
```

Reply:
I'm wondering if @deads2k  was suggesting that the entire placement group policy should be a discriminating union?

Reply:
That example makes sense to me, but I would suggest breaking it out into a separate type, which could be inlined:
```go
type Placement struct {
  // ...
  GroupUnion  `json:",inline"`
}

// +union
type GroupUnion struct {
  // +unionDiscriminator
  // +kubebuilder:validation:Required
  // +required
  GroupType AWSPlacementGroupType

  // +optional
  Cluster *AWSClusterPlacement

  // +optional
  Partition *AWSPartitionPlacement

  // +optional
  Spread *AWSSpreadPlacement  
}

type AWSClusterPlacement struct {}

type AWSPartitionPlacement struct {
  Count int32
  Number int32
}

type AWSSpreadPlacement struct {}
```

Reply:
I can do this yep, though, I'm not hugely familiar with the union types, does this mean users have to set one of the structs to be non-empty? Or can they just set the type for Cluster or Spread and omit the struct? Also, do we even need to add those structs to the API if they are empty/not required right now?

Reply:
The user can specify the discriminator, or specify one of the structs, or both.  I think you can omit the empty structs and corresponding fields if you prefer.